### ISSUE TITLE
Task arguments act like inputs

### ISSUE THREAD

**Issue by orius123** [CloudSlang/cloud-slang#288]
for example they perform the required logic and in cast that a required input is not available fail the flow instead of passing the value as-is

**Comment by Bonczidai** [comment]
Let's decide the syntax and behavior we want, so ..

Task arguments will not behave (in general) like flow / operation inputs.
Task arguments will have a name and a default value (expression) which will be evaluated and passed to the task (similar behavior to `overridable: false`). They will **not** support the following input modifiers: system property, overridable, encrypted.

Need to decide what should happen in case of the `required` tag. I think that here we should either support the tag or make `required: false` the default - since it is common use case to pass optional values all over the flows

When naming should we go with `task inputs` or `task arguments` (other suggestion) ?

@meirwah @orius123 WDYT on this?

**Comment by orius123** [comment]
I think that task arguments should act like function arguments in python. and come up with oneliner.
for example `containers.pull_container: host = 'localhost', port = 8080`
or even `containers.pull_container(host = 'localhost', port = 8080)`
Regarding `required` I think it is the operation/sub-flow responsibility

**Comment by Bonczidai** [comment]
also need to consider the large number of task arguments & we should align to yaml syntax - so here are a couple of suggestions..

What do you prefer? Suggestions?

`#1`
```yaml
    - get_containers_in_cluster:
        do:
          containers.get_filtered_containers(docker_options, all_containers = true, excluded_images = excluded, host = 'localhost', port, username, password, private_key_file, character_set, pty, timeout = 300000, close_session, agent_forwarding)
        publish:
          - container_ids
```

`#2`
```yaml
    - get_containers_in_cluster:
        do:
          containers.get_filtered_containers: >
            docker_options, all_containers = true, excluded_images = excluded, host = 'localhost', port,
            username, password, private_key_file, character_set, pty, timeout = 300000, close_session, agent_forwarding
        publish:
          - container_ids
```

`#3`
```yaml
    - get_containers_in_cluster:
        do:
          containers.get_filtered_containers:
            - docker_options
            - all_containers: true
            - excluded_images: excluded
            - host: 'localhost'
            - port
            - username
            - password
            - private_key_file
            - character_set
            - pty
            - timeout: 300000
            - close_session
            - agent_forwarding
        publish:
          - container_ids
```

**Comment by meirwah** [comment]
I will go with #3, 
I do not like the oneliner as #1

**Comment by Bonczidai** [comment]
also here is a brief list of planned changes affecting the codebase:

### Entities:    

- new type: **Argument** extends InOutParam
	- has name, expression

### Compiler

- Task will have a `List<Argument>` instead of a `List<Input>`
- add proper transformer
- treat case accordingly when building the executable (precompile phase) and the execution plan (compile phase)

### Runtime

- add ArgumentsBinding
- modify TaskSteps accordingly

basically is around introducing a new Type called `Argument (?)` for task inputs - WDYT?

**Comment by orius123** [comment]
I agree with Meir that option 3 is best when you have a lot of outputs. 
But when you have few one line is much better. 
We can write one line by just using yank one line list, But I like 1 and 2 better. 
Anyway we should support and document both.

**Comment by Bonczidai** [comment]
I opened a PR - currently supports syntax `3`. I will continue to add features and update the tests on this branch.

Some important aspects:
- it will support both on-liner (`1`/`2` ?) and list syntax
- default behavior: `required: false` (executable will validate inputs)
- task has inputs called **arguments**

In the meantime you can add comments - @meirwah @orius123

**Comment by orius123** [comment]
Just to make clear.
The behavior is `overridable: false` by default?
I really like option 1 :trophy: because I think it emphasize the way of how we see tasks as function calls rather than inputs declaration, but I can live with 2..

**Comment by Bonczidai** [comment]
ok - will start with option 1 (need to check the refactoring effort it requires)

.. `overridable: false` is the default behavior

**Comment by orius123** [comment]
@meirwah 
R U good with `option 1(look = on, what = I, wrote= previously)`
O do U prefer `option 2: that = looks, more = like, {ansibe, mistral}`

**Comment by meirwah** [comment]
#2

**Comment by orius123** [comment]
You heard the FA

**Comment by meirwah** [comment]
On what this is stuck? @oritstone @Bonczidai ?

**Comment by Bonczidai** [comment]
we can merge this -- we should also make a release so we can use it in the build system
content side is almost ready (CloudSlang/cloud-slang-content#330)
